File: 23.合并 K 个升序链表.py
#
# @lc app=leetcode.cn id=23 lang=python3
# @lcpr version=30204
#
# [23] 合并 K 个升序链表
#


# @lcpr-template-start

# @lcpr-template-end
# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def mergeKLists(self, lists: List[Optional[ListNode]]) -> Optional[ListNode]:
        if not lists:
            return None
        
        size = len(lists)
        return self.merge_sort(lists, 0, size - 1)
    
    def merge_sort(self, lists: List[ListNode], left: int, right: int) -> ListNode:
        if left == right:
            return lists[left]
        
        mid = left + (right - left) // 2
        left_node = self.merge_sort(lists, left, mid)
        right_node = self.merge_sort(lists, mid+1, right)
        return self.merge(left_node, right_node)

    def merge(self, left: List[ListNode], right: List[ListNode]):
        root = ListNode(-1)
        cur = root
        while left and right:
            if left.val < right.val:
                cur.next = left
                left = left.next
            else:
                cur.next = right
                right = right.next
            cur = cur.next
        if left:
            cur.next = left

        if right:
            cur.next = right

        return root.next
        
# @lc code=end



#
# @lcpr case=start
# [[1,4,5],[1,3,4],[2,6]]\n
# @lcpr case=end

# @lcpr case=start
# []\n
# @lcpr case=end

# @lcpr case=start
# [[]]\n
# @lcpr case=end

#

File: 23.合并 K 个升序链表.rs
/*
 * @lc app=leetcode.cn id=23 lang=rust
 * @lcpr version=30204
 *
 * [23] 合并 K 个升序链表
 */

// @lcpr-template-start

// @lcpr-template-end
// @lc code=start
// Definition for singly-linked list.
// #[derive(PartialEq, Eq, Clone, Debug)]
// pub struct ListNode {
//   pub val: i32,
//   pub next: Option<Box<ListNode>>
// }
//
// impl ListNode {
//   #[inline]
//   fn new(val: i32) -> Self {
//     ListNode {
//       next: None,
//       val
//     }
//   }
// }
impl Solution {
    pub fn merge_k_lists(mut lists: Vec<Option<Box<ListNode>>>) -> Option<Box<ListNode>> {
        if lists.len() <= 1 {
            return lists.pop()?;
        } else if lists.len() == 2 {
            return Self::merge_two_lists(lists.pop().unwrap(), lists.pop().unwrap());
        }

        let right_arr = lists.split_off(lists.len() >> 1);
        let left = Self::merge_k_lists(lists);
        let right = Self::merge_k_lists(right_arr);

        return Self::merge_k_lists(vec![left, right]);
    }

    pub fn merge_two_lists(
        mut list1: Option<Box<ListNode>>,
        mut list2: Option<Box<ListNode>>,
    ) -> Option<Box<ListNode>> {
        let mut dummy = ListNode::new(0);
        let mut ptr = &mut dummy;

        while let (Some(n1), Some(n2)) = (list1.as_ref(), list2.as_ref()) {
            if n1.val < n2.val {
                ptr.next = list1;
                ptr = ptr.next.as_mut().unwrap();
                list1 = ptr.next.take();
            } else {
                ptr.next = list2;
                ptr = ptr.next.as_mut().unwrap();
                list2 = ptr.next.take();
            }
        }
        ptr.next = list1.or(list2);
        dummy.next
    }
}
// @lc code=end

/*
// @lcpr case=start
// [[1,4,5],[1,3,4],[2,6]]\n
// @lcpr case=end

// @lcpr case=start
// []\n
// @lcpr case=end

// @lcpr case=start
// [[]]\n
// @lcpr case=end

 */
